keps/sig-node/draft-20180702-pod-restartpolicy-to-restart-whole-pod.md

```diff
@@ -0,0 +1,125 @@
+reviewers:
+  - "@mtaufen"
+  - "@smarterclayton"
+approvers:
+  - "@liggitt"
+  - "@derekwaynecarr"
+editor:
+creation-date: 2018-07-04
+last-updated: 2018-07-04
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# New pod restartPolicy to restart the whole pod instead of just a container
+
+## Table of Contents
+* [New pod restartPolicy to restart the whole pod instead of just a container](#new-pod-restartpolicy-to-restart-the-whole-pod-instead-of-just-a-container)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non\-Goals](#non-goals)
+  * [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Recreate Pod Sandbox](#recreate-pod-sandbox)
+        * [Pros](#pros)
+        * [Cons](#cons)
+      * [Re\-use Pod Sandbox](#re-use-pod-sandbox)
+        * [Pros](#pros-1)
+        * [Cons](#cons-1)
+    * [Risks and Mitigations](#risks-and-mitigations)
+  * [Graduation Criteria](#graduation-criteria)
+  * [Implementation History](#implementation-history)
+  * [Alternatives](#alternatives)
+
+## Summary
+
+In a pod with multiple containers, if one of the containers terminates and if the restart policy mandates that it be restarted (as in [`restartPolicy`][rp] of `OnFailure` or `Always`) then the kubelet restarts only the terminated container. The other containers are left as they are.
+
+This behaviour, while generally desirable, makes some scenarios such as the ones involving [initContainers][issue] or some complex interaction between the containers of the pod cumbersome to implement.
+
+To address such scenarios, this proposal introduces a new pod `restartPolicy` called `AlwaysPod` to make it possible to restart the whole pod (including the initContainers) whenever the `restartPolicy` `Always` would have restarted just one of the containers of the pod.
+
+## Motivation
+
+The `OnFailure` and `Always` [restart polices][rp] efficiently manage the life-cycle of the containers of a pod. The support for multiple containers in a pod also enable better modularity and separation of concerns between different containers. It also promotes looser coupling between components.
+
+[Init containers] provide some additional support for modularity and looser coupling for the functionality of initialization of the pod. They make it possible to separate the initialization from the rest of the pod to enhance both modularity as well as security.
+
+But both the `OnFailure` as well as the `Always` restart policies restart the individual containers in question and not the whole pod. This is, for the most part, desirable, even optimal.
+
+However, there are scenarios (some documented in [this issue][issue]) where the many containers in the pod (including init containers) might be interlinked or inter-dependent in such a way as to require closer co-ordination when any one of its containers are restarted.
```

Comment:
@amshuman-kr We discussed this in SIG Apps this week. The motivation for adding this was not completely understood. Could you please expand on user stories ([see the template](https://github.com/kubernetes/community/blob/master/keps/0000-kep-template.md)) that outline situations that this would be used for. This will help others understand the need and now this can help with it.